### DISCUSSION TITLE
Enhancement idea, and how to build?

### DISCUSSION THREAD

**Discussion by mjh65** [mauer/xmidictrl#19] [Ideas]
Hi Marco,
I'm using an X-Touch with a custom xmini2vjoy script and a vJoy virtual joystick. The setup is a little unresponsive to rotating knobs, and I think your X-Plane plugin will be a better solution.
I'm also flying in VR, so I like the idea of seeing visual feedback in the HMD, since I'm operating the controls by feel/instinct!
In my current setup I have configured the button press of the rotating knobs to act as an 'inner/outer' mode toggle (rather than sending a command). I use this so that a single rotating knob can control the tuning of a radio - rotate for coarse frequency, press, rotate for fine tuning, etc.
I'd like to try adding this capability to your plugin. Could you post instructions on how to build the plugin from source, and if you have any suggestions for a preferred approach to my proposal please let me know.
Thanks,
Mike

**Comment by mauer** [comment]
Hi,

I'm sorry for the late reply. The plugin can actually already handle that - I just haven't documented it well enough ([link](https://mauer.github.io/xmidictrl/#/sublayers)). It's called sublayers and allows you to add multiple mappings to a single button or encoder.

I use this feature in the [B742](https://github.com/mauer/xmidictrl/blob/main/resources/examples/B742_XMidiCtrl.toml) example.

First of all, you have to declare a sublayer dataref. For convenience I already create a default one called **xmidictrl/sublayer**:

`title      = "Settings for the Felis 742"
version    = "1.00"
sl_dataref = "xmidictrl/sublayer"`

Next would be to add the possible sl_dataref values to your mappings:

`  # Encoders
  { ch = 11, cc = 1, sl = "0", type = "enc", command_up = "sim/instruments/barometer_up", command_down = "sim/instruments/barometer_down" },
  { ch = 11, cc = 1, sl = "1", type = "enc", command_up = "sim/instruments/barometer_stby_up", command_down = "sim/instruments/barometer_stby_down" },`

This means, that the first mapping will be executes, when the sl_dataref value is equal to 0 and the second mapping when the value is equal to 1.

Now we need an option to toggle the dataref between 0 and 1. This option should be triggered when pressing the toggle button. Lets add a mapping for that (XMidiCtrl already ships a command for that):

`  # Encoder Buttons
  { ch = 11, cc = 9,  type = "cmd", command = "xmidictrl/toggle_sublayer" }`

I hope my explanation helps and I will try to write proper documentation for it.

**Reply by mjh65** [reply]
Hi Marco,

Thanks for the explanation and examples.

I recall reading about sublayers, but I think I'd interpreted them as relating to some predefined state of the aircraft. I hadn't understood that it was possible to create a new dataref just to hold the sublayer state.

So I have a couple of further questions (but I will also experiment to try and find the answers):
1) is it possible to define many sublayers, ie one for each rotary?
2) is it possible for a sublayer to have more than 2 states (I'm thinking of the option to (eg) overload COM1 coarse/fine and COM2 coarse/fine on a single rotary!)?

Thanks,
Mike

**Comment by mjh65** [comment]
BTW, I figured out how to build the plugin, and I made a few updates so that it is possible to build using Visual Studio 2022. This was useful for me since I prefer the VS built-in debugger. I've submitted a PR to include this in your repo if you don't object.

**Comment by mjh65** [comment]
Some further ideas. So far I think that the sublayer feature doesn't quite do what I was thinking of. I'd like to do a bit more experimenting and debugging, but I find this is quite a heavyweight process running the plugin inside X-Plane.

It would be good to be able to build the core of XMidiCtrl as a library and then link this into the X-Plane plugin, or into a simple executable for standalone testing.

Another alternative might be to create an application that loads the plugin as a DLL, but I think there might be quite a lot of XP interfaces that would need to be implemented.

Do you have any thoughts?
